test/integration/pause-resume-test.js

```diff
@@ -0,0 +1,246 @@
+    test.ok(!packetTransformRs.flowing,
+      'Packet transform is not paused.');
+    test.ok(packetTransformWs.length <= packetTransformWs.highWaterMark && packetTransformRs.length <= packetTransformRs.highWaterMark,
+      'Packet transform has large amount of data buffered.');
+    const tokenTransformWs = connection.tokenStreamParser.parser._writableState;
+    const tokenTransformRs = connection.tokenStreamParser.parser._readableState;
+    test.ok(!tokenTransformRs.flowing,
+      'Token transform is not paused.');
+    test.ok(tokenTransformWs.length <= tokenTransformWs.highWaterMark,
+      'Token transform input buffer overflow.');
+    test.ok(tokenTransformRs.length < packetSize / 3,
+      'Token transform output buffer has large amount of data buffered.');
+  }
+
+  function fail(msg) {
+    if (failed) {
+      return; }
+    failed = true;
+    test.ok(false, msg);
+    connection.close();
+  }
+};
+
+// This test reads only a few rows and makes a short pause after each row.
+// The test verifies that:
+//  - Pause/resume works correctly when applied after the last packet of a TDS
+//    message has already been dispatched by MessageIO.ReadablePacketStream.
+//    (EOM / packet.isLast() has already been detected when pause() is called.
+//    The 'message' event emitted by MessageIO has to be channeled through
+//    the token parser transform.)
+//  - The data flow is automatically resumed when a query is canceled.
+exports.testTransitions = function(test) {
+  const totalRequests = 3;
+  const rowsPerRequest = 4;
+  const delayTime = 100;                                   // pause delay time in ms
+  const requestToCancel = 2;                               // 1-based position of request to be canceled
+  const rowToCancel = 2;                                   // 1-based position of row at which connection.cancel() will be called
+  const connection = this.connection;
+  let request;
+  let requestCount = 0;
+  let rowCount;
+  let paused = false;
+  let canceled = false;
+
+  connection.on('error', function(err) {
+    test.ifError(err);
+  });
+  openRequest();
+
+  function openRequest() {
+    let sql = 'select 1';
+    for (let i = 2; i <= rowsPerRequest; i++) {
+      sql = sql + ' union all select ' + i;
+    }
+    request = new Request(sql, onRequestCompletion);
+    request.on('row', processRow);
+    rowCount = 0;
+    paused = false;
+    canceled = false;
+    connection.execSql(request);
+  }
+
+  function onRequestCompletion(err) {
+    requestCount++;
+    if (requestCount == requestToCancel) {
+      test.ok(err && err.code == 'ECANCEL');
+      test.equal(rowCount, rowToCancel);
+    } else {
+      test.ifError(err);
+      test.equal(rowCount, rowsPerRequest);
+    }
+    if (requestCount < totalRequests) {
+      openRequest();
+    } else {
+      test.done();
+    }
+  }
+
+  function processRow(columns) {
+    if (canceled) {
```

Comment:
Agreed, this is a better approach. This prevents the possibility of an application _bug_ where it chokes on 'row' events coming in after 'cancel'. Making it difficult to write bugs is good :-)